README.md
easy-mock-api
=============

easy!

[![oclif](https://img.shields.io/badge/cli-oclif-brightgreen.svg)](https://oclif.io)
[![Version](https://img.shields.io/npm/v/easy-mock-api.svg)](https://npmjs.org/package/easy-mock-api)
[![Downloads/week](https://img.shields.io/npm/dw/easy-mock-api.svg)](https://npmjs.org/package/easy-mock-api)
[![License](https://img.shields.io/npm/l/easy-mock-api.svg)](https://github.com/mihirogi/easy-mock-api/blob/master/package.json)


* [Usage](#usage)
* [Commands](#commands)

# Usage

```bash
$ npx easy-mock-api init // generate json file
$ npx easy-mock-api start // start server
```

# Commands

*  [`easy-mock-api init`](#init)
*  [`easy-mock-api start`](#start)

## `init`

describe the command here

```
USAGE
  $ npx easy-mock-api init
```

_See code: [src/commands/init.ts](https://github.com/mihirogi/easy-mock-api/blob/v0.0.0/src/commands/init.ts)_

## `start`

describe the command here

```
USAGE
  $ npx easy-mock-api start
```

_See code: [src/commands/init.ts](https://github.com/mihirogi/easy-mock-api/blob/v0.0.0/src/commands/init.ts)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.2](https://github.com/mihirogi/easy-mock-api/compare/v0.0.1...v0.0.2) (2020-03-23)


### Bug Fixes

* path ([#7](https://github.com/mihirogi/easy-mock-api/issues/7)) ([#8](https://github.com/mihirogi/easy-mock-api/issues/8)) ([27bd810](https://github.com/mihirogi/easy-mock-api/commit/27bd810185b69f139dcdf041dba323fde128b084))

### 0.0.1 (2020-03-23)


### Features

* install commitlint and stanard-version ([#6](https://github.com/mihirogi/easy-mock-api/issues/6)) ([30cbb8b](https://github.com/mihirogi/easy-mock-api/commit/30cbb8bdbc9c775be562367518aa3490efbdec0b))
* start command ([#3](https://github.com/mihirogi/easy-mock-api/issues/3)) ([fc73095](https://github.com/mihirogi/easy-mock-api/commit/fc730954001c44d069b230daa96aee1cc4554f26))


### Bug Fixes

* refactor function ([#5](https://github.com/mihirogi/easy-mock-api/issues/5)) ([520b540](https://github.com/mihirogi/easy-mock-api/commit/520b540ef0085fa1295b0d104424acbabd20d9e8))


test/commands/init.test.ts
import {expect, test} from '@oclif/test'
import enquirer = require('enquirer');
import {Response, DefaultJsonFormat} from '../../src/commands/init'
import * as fs from 'fs'
import * as config from '../../src/constants/config'

describe('init', () => {
  test
  .stub(enquirer, 'prompt', () => (): Response => {
    return {value: '/hello'}
  })
  .stdout({print: true})
  .command('init')
  .it('runs init', () => {
    expect(fs.existsSync(`${config.dirPath}/${config.fileName}`)).is.true
    const createdJson: DefaultJsonFormat = require(`${config.dirPath}/${config.fileName}`)
    expect(createdJson.port).to.have.equal(4200)
    expect(createdJson.data).to.have.keys(['/hello', '/sample'])
    Object.keys(createdJson.data).forEach(key => {
      expect(createdJson.data[key]).to.have.keys(['body', 'delay', 'statusCode'])
    })
  })
})


test/commands/start.test.ts
import {test, expect} from '@oclif/test'
import ApiServer from '../../src/singletons/api-server'
import * as request from 'request-promise-native'

describe('start', () => {
  test
  .stub('loadDefinedApiMapFile', () => ({
    port: 4200,
    data: {
      '/hello': {
        body: {message: 'please rewrite with JSON format data'},
        delay: 5000,
        statusCode: 200,
      },
      '/sample': {
        body: {message: '/sample is example data'},
        delay: 2000,
        statusCode: 200,
      },
    },
  }))
  .stdout({print: true})
  .command('start')
  .it('runs start', async () => {
    const response = await request.get('http://localhost:4200/sample')
    expect(JSON.parse(response)).have.to.key('message')
    expect(JSON.parse(response).message).have.to.contain('/sample is example data')
  })
}).afterEach(() => ApiServer.close())


src/commands/init.ts
import {Command, flags} from '@oclif/command'
import enquirer = require('enquirer');
import {dirPath, fileName} from '../constants/config'
import {generateJson, outputJson, createDir} from '../functions/api'

export default class Init extends Command {
  static description = 'generate json file';

  static flags = {
    help: flags.help({char: 'h'}),
  };

  async run() {
    const response: Response = await enquirer.prompt({
      type: 'input',
      name: 'value',
      message: 'enter the api path',
      initial: '/sample',
    })
    createDir(dirPath)
    outputJson(
      `${dirPath}/${fileName}`,
      JSON.stringify(generateJson(response.value))
    )
  }
}

export interface Response {
  value: string;
}
export interface DefaultJsonFormat {
  port: number;
  data: {
    [key: string]: {
      body: any;
      delay: number;
      statusCode: number;
    };
  };
}


src/commands/hello.ts
import { Command, flags } from '@oclif/command';

export default class Hello extends Command {
  static description = 'describe the command here';

  static examples = [
    `$ easy-mock-api hello
hello world from ./src/hello.ts!
`
  ];

  static flags = {
    help: flags.help({ char: 'h' }),
    // flag with a value (-n, --name=VALUE)
    name: flags.string({ char: 'n', description: 'name to print' }),
    // flag with no value (-f, --force)
    force: flags.boolean({ char: 'f' })
  };

  static args = [{ name: 'file' }];

  async run() {
    const { args, flags } = this.parse(Hello);

    const name = flags.name || 'world';
    this.log(`hello ${name} from ./src/commands/hello.ts`);
    if (args.file && flags.force) {
      this.log(`you input --force and --file: ${args.file}`);
    }
  }
}


src/commands/start.ts
import {Command} from '@oclif/command'
import {IncomingMessage, ServerResponse} from 'http'
import ApiServer from '../singletons/api-server'
import {generateApiMap} from '../functions/api'

export default class Start extends Command {
  static description = 'start mock api server';

  async run() {
    const apiMap = generateApiMap()
    ApiServer.createServer(
      (request: IncomingMessage, response: ServerResponse) => {
        this.log(`Request to ${request.url}`)
        // レスポンス定義の中にリクエストされたAPIの定義がなければ、_undefinedとする
        const url =
          (apiMap.data[request.url as string] && request.url) || '_undefined'
        setTimeout(() => {
          response.writeHead(apiMap.data[url].statusCode, {
            'Content-type': 'application/json',
          })
          response.write(JSON.stringify(apiMap.data[url].body))
          response.end()
          request.on('end', () => request.connection.end())
        }, apiMap.data[url].delay)
      }
    )
    ApiServer.listen(apiMap.port, () => {
      this.log(`Mock Server on port ${apiMap.port}`)
    })
  }
}


src/functions/api.ts
import {DefaultJsonFormat} from '../commands/init'
import {dirPath, fileName} from '../constants/config'
import * as fs from 'fs'
import mkdirp = require('mkdirp');

export const loadDefinedApiMapFile = (path: string): DefaultJsonFormat => {
  return require(path)
}

export const generateApiMap = (): DefaultJsonFormat => {
  const defineApiMap = loadDefinedApiMapFile(`${dirPath}/${fileName}`)
  return {
    ...defineApiMap,
    data: {
      ...defineApiMap.data,
      _undefined: {
        // 定義されてないAPIのレスポンス用
        body: {message: 'undefined api'},
        delay: 0,
        statusCode: 200,
      },
    },
  }
}

export const generateJson = (endpoint: string): DefaultJsonFormat => {
  return {
    port: 4200,
    data: {
      [endpoint]: {
        body: {
          message: 'please rewrite with JSON format data',
        },
        delay: 1000,
        statusCode: 200,
      },
      '/sample': {
        body: {
          message: '/sample is example data',
        },
        delay: 1000,
        statusCode: 200,
      },
    },
  }
}

export const outputJson = (path: string, json: string) => {
  fs.writeFileSync(path, json)
}

export const createDir = (path: string) => {
  mkdirp.sync(path)
}
